### ISSUE TITLE
No protection from infinite structural loops

### ISSUE THREAD

**Issue by dantleech** [phpactor/phpactor#1627]
it's possible to create an infinite loop if a class-like structure creates a cycle:

```
<?php

class Foobar extends One {
    public function bar(): string;
}

class One extends Foobar {
    use TraitOne;
}

$one = new One();
```
```
<?php

trait TraitTwo {
    use TraitOne;
    public function onTraitTwo(): string {}
    public function onTraitThree(): int {}
}

trait TraitOne {
    use TraitTwo;

    public function onTraitOne(): string {}
}

class One {
    use TraitOne;
}

$one = new One();

wrAssertType('string', $one->onTraitOne());
wrAssertType('string', $one->onTraitTwo());
wrAssertType('int', $one->onTraitThree());
```

probably need to refactor to resolve the hierarchy before traversing it. Possibly this also improves perfomance in some cases.

**Comment by dantleech** [comment]
this is difficult due to WRs internals. each reflection class-like will attempt to aggregate it's own methods, properties and constants, which means that we'd need to track the "visited" classes in multiple places and muddy the API, and the internals are already pretty messy and inconsistent. A short-term fix might be to have a shared cache to track visited classes, but I couldn't get that to work due to the order in which things are resolved.

the best solution would be to solve the classes to traverse up-front and then iterate over them and aggregate the members, for this there would need to be a  method which returned only the classes member(s) and a method to return the "member providing" class-likes in order of precedence (e.g. all traits in the inheritence hierarchy, all interfaces, all parent classes).

it would probably be OK to have a single `ownMembers(): ReflectionMemberCollection` as the source of truth rather than `ownProperties`, `ownConstants`, `ownCases`, `ownMethods` as this method would need to be propagated publically over all class-likes, and it's really an internal detail (as we can already do `$members->ownedBy(Foo::class)`).

**Comment by dantleech** [comment]
`ownMembers` is fine, but would also need `ownCollaborators`,

```php

function resolveCollaborators(ReflectionClassLike $classLike, $resolved = [])
{
    foreach ($classLike->collaborators() as $collaborator) {
        if (isset($collaboratorsp[$collaborator->name()])) {
            continue;
        }
        $resolved[$collaborator->name()] = $collaborator;
        resolveCollaborators($collaborator, $resolved);
    }
    return $resolved;
}
```

**Comment by PierreRambaud** [comment]
That's maybe the bug I have :sweat_smile:

**Comment by dantleech** [comment]
This _should_ be fixed as of now #1737 would be good to know if it addresses your issue @pierredup

**Comment by dantleech** [comment]
ok _really_ fixed in #1821 (mixins too)
